roles/tests-integration/tests/common/sniffer.rs

```diff
@@ -0,0 +1,380 @@
+use async_channel::{Receiver, Sender};
+use codec_sv2::{
+    framing_sv2::framing::Frame, HandshakeRole, Initiator, Responder, StandardEitherFrame,
+};
+use key_utils::{Secp256k1PublicKey, Secp256k1SecretKey};
+use network_helpers_sv2::noise_connection_tokio::Connection;
+use roles_logic_sv2::{
+    parsers::{
+        CommonMessages,
+        JobDeclaration::{
+            AllocateMiningJobToken, AllocateMiningJobTokenSuccess, DeclareMiningJob,
+            DeclareMiningJobError, DeclareMiningJobSuccess, IdentifyTransactions,
+            IdentifyTransactionsSuccess, ProvideMissingTransactions,
+            ProvideMissingTransactionsSuccess, SubmitSolution,
+        },
+        PoolMessages,
+        TemplateDistribution::{self, CoinbaseOutputDataSize},
+    },
+    utils::Mutex,
+};
+use std::{collections::VecDeque, net::SocketAddr, sync::Arc};
+use tokio::{
+    net::{TcpListener, TcpStream},
+    select,
+};
+
+type MessageFrame = StandardEitherFrame<PoolMessages<'static>>;
+type MsgType = u8;
+
+#[derive(Debug, PartialEq)]
+pub enum SnifferError {
+    DownstreamClosed,
+    UpstreamClosed,
+}
+
+#[derive(Debug, Clone)]
+pub struct Sniffer {
+    downstream: SocketAddr,
+    upstream: SocketAddr,
+    downstream_messages: MessagesAggregator,
+    upstream_messages: MessagesAggregator,
+}
+
+impl Sniffer {
+    pub async fn new(downstream: SocketAddr, upstream: SocketAddr) -> Self {
+        Self {
+            downstream,
+            upstream,
+            downstream_messages: MessagesAggregator::new(),
+            upstream_messages: MessagesAggregator::new(),
+        }
+    }
+
+    pub async fn start(self) {
+        let (downstream_receiver, downstream_sender) =
+            Self::create_downstream(Self::wait_for_client(self.downstream).await)
+                .await
+                .expect("Failed to create downstream");
+        let (upstream_receiver, upstream_sender) = Self::create_upstream(
+            TcpStream::connect(self.upstream)
+                .await
+                .expect("Failed to connect to upstream"),
+        )
+        .await
+        .expect("Failed to create upstream");
+        let downstream_messages = self.downstream_messages.clone();
+        let upstream_messages = self.upstream_messages.clone();
+        let _ = select! {
+            r = Self::recv_from_down_send_to_up(downstream_receiver, upstream_sender, downstream_messages) => r,
+            r = Self::recv_from_up_send_to_down(upstream_receiver, downstream_sender, upstream_messages) => r,
+        };
+    }
+
+    pub fn downstream_state(&self) -> Option<PoolMessages<'static>> {
+        self.downstream_messages.current_state()
+    }
+
+    pub fn upstream_state(&self) -> Option<PoolMessages<'static>> {
+        self.upstream_messages.current_state()
+    }
```

Comment:
the naming here is confusing

maybe better names could be:
- `pop_downstream_message()`
- `pop_upstream_message()`